### PR TITLE
Feature 6987 implement sarext

### PR DESCRIPTION
**Pull request by nbkell**
Implemented SAREXT indicator. 

#### Description
See discussion in: [6987](https://github.com/QuantConnect/Lean/issues/6987)

#### Related Issue
[6987](https://github.com/QuantConnect/Lean/issues/6987)

#### Motivation and Context
Adds indicator requested in [6987](https://github.com/QuantConnect/Lean/issues/6987)

#### Requires Documentation Change
None that I know of. 

#### How Has This Been Tested?
Tests have been added. See: [6987](https://github.com/QuantConnect/Lean/issues/6987)

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Tests/Indicators/ParabolicStopAndReverseTestsExtended.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            var sarext = new ParabolicStopAndReverseExtended(sarStart : 100m, offsetOnReverse : 0.02m, 
+                afStartShort : 0.03m, afIncrementShort : 0.02m, afMaxShort : 0.4m, 
+                afStartLong : 0.02m, afIncrementLong : 0.01m, afMaxLong : 0.3m);
+            TestHelper.TestIndicator(
+                sarext,
+                TestFileName,
+                "SAREXT_PARAM4",
+                (ind, expected) => Assert.AreEqual(expected, (double)sarext.Current.Value, delta: 1e-4)
+            );
+        }
```
Comment (jhonabreul) [inline Tests/Indicators/ParabolicStopAndReverseTestsExtended.cs:90]: All these tests are the same with different parameters, please use the `TestCase` attribute to parametrize a single test.

**Review by jhonabreul — CHANGES_REQUESTED** [review]
Nice, thank you! Leaving a few comments.

Also, is there a way we can reuse this class as a base for `ParabolicStopAndReverse` so we don't duplicated code? `ParabolicStopAndReverse` could call the constructor of this class passing the right parameters that would make it behave exactly as it is now.

**Review by jhonabreul — CHANGES_REQUESTED** [review]
Looking good!

- Remember to use the test data file in https://github.com/QuantConnect/Lean/issues/6987#issuecomment-1777044713
- What about https://github.com/QuantConnect/Lean/pull/7528#issue-1957958591?

File: .DS_Store
[the remark this answers is not in the document]
Comment (jhonabreul) [reply .DS_Store:1]: Can add it to gitignore!

File: Indicators/ParabolicStopAndReverseExtended.cs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        {
+            // On first iteration we can’t produce an SAR value so we save the current bar and return zero
+            if (Samples == 1){
+
+                 _previousBar = input;
+                // Makes sense to return _sarInit when its non-zero
+                if(_sarInit > 0)
+                    return _sarInit; 
+                else if(_sarInit < 0)
+                    return Math.Abs(_sarInit);
```
[the remark this answers is not in the document]
Comment (nbkell) [reply Indicators/ParabolicStopAndReverseExtended.cs]: Fixed

File: Indicators/ParabolicStopAndReverseExtended.cs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+        }
+
+        /// <summary>
+        /// Returns true if DM > 0 between today and yesterday's tradebar (false otherwise)
+        /// </summary>
+        private bool HasNegativeDM(IBaseDataBar currentBar){
+            if (currentBar.Low >= _previousBar.Low)
+                return false;
+            decimal diff1 = currentBar.High - _previousBar.High; 
+            decimal diff2 = _previousBar.Low - currentBar.Low;
```
[the remark this answers is not in the document]
Comment (nbkell) [reply Indicators/ParabolicStopAndReverseExtended.cs]: Fixed.

File: Tests/Indicators/ParabolicStopAndReverseTestsExtended.cs (hunk 1 above)
Comment (nbkell) [reply Tests/Indicators/ParabolicStopAndReverseTestsExtended.cs]: Fixed.

File: Tests/TestData/.DS_Store
[the remark this answers is not in the document]
Comment (nbkell) [reply Tests/TestData/.DS_Store:1]: Apologies! I forgot to set PR to draft before pushing; didn't intend for you to review yet :)

File: Indicators/ParabolicStopAndReverseExtended.cs [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+           }
+           // same set up as standard PSAR when _sarInit = 0 
+           else
+           {
+                _isLong = !HasNegativeDM(currentBar); 
+                _sar = _isLong ? _previousBar.Low : _previousBar.High;
+           }
+
+            // initialize extreme point 
+            _ep = _isLong ? currentBar.High : currentBar.Low;
```
[the remark this answers is not in the document]
Comment (nbkell) [reply Indicators/ParabolicStopAndReverseExtended.cs:166]: This was done to be consistent with talib's implementation, which sets <code>ep</code>only based on the second bar and ignores the first. For the reference you can find talib's implementation [here](https://github.com/TA-Lib/ta-lib/blob/main/src/ta_func/ta_SAREXT.c). The code for initializing ep is on lines 512-537. (Note that during initialization <code>todayIndex</code> is set to 1, i.e., second bar.)

**Comment by nbkell** [comment]
Many thanks for your comments and quick review! I've incorporated your changes including an updated implementation of PSAR that uses SAREXT's code. 

Remaining questions regarding testing:

1. I'm a bit confused on how @LouisSzeto test file is being generated, since when I run the same script locally I end up with a different .csv. For reference, I've posted below the script I'm using to generate the current version of <code>spy_sarext.txt</code>.

2. As outlined in the second half on my comment in [6987](https://github.com/QuantConnect/Lean/issues/6987), Lean's current implementation of PSAR differs slightly from that of talib. Thus the updated PSAR implementation using SAREXT's code actually fails the current test for PSAR. I've updated the test to verify it matches talib's SAR function (it's an additional column in <code>spy_extsar.txt</code> generated in script below), but ultimately let me know how this should be resolved. 

Test File Script:


```
import pandas as pd
from talib import SAREXT, SAR

df = pd.read_csv('https://github.com/QuantConnect/Lean/raw/master/Data/equity/usa/daily/spy.zip',
                          index_col=0, names=['date', "open", "high", "low", "close", "volume"])
df = df/10000
df = df.reset_index().rename(columns={'index':'date'})

df['SAREXT'] = SAREXT(low=df.low, high=df.high)
df['SAR'] = SAR(low=df.low, high=df.high)

df['SAREXT_PARAM1'] = SAREXT(low=df.low, 
                             high=df.high, 
                             startvalue=0, 
                             offsetonreverse=0, 
                             accelerationinitlong=0.02, 
                             accelerationlong=0.01, 
                             accelerationmaxlong=0.3, 
                             accelerationinitshort=0.03, 
                             accelerationshort=0.02, 
                             accelerationmaxshort=0.4)

df['SAREXT_PARAM2'] = SAREXT(low=df.low, 
                             high=df.high, 
                             startvalue=100, 
                             offsetonreverse=0, 
                             accelerationinitlong=0.02, 
                             accelerationlong=0.01, 
                             accelerationmaxlong=0.3, 
                             accelerationinitshort=0.03, 
                             accelerationshort=0.02, 
                             accelerationmaxshort=0.4)

df['SAREXT_PARAM3'] = SAREXT(low=df.low, 
                             high=df.high, 
                             startvalue= -95, 
                             offsetonreverse=0, 
                             accelerationinitlong=0.02, 
                             accelerationlong=0.01, 
                             accelerationmaxlong=0.3, 
                             accelerationinitshort=0.03, 
                             accelerationshort=0.02, 
                             accelerationmaxshort=0.4)

df['SAREXT_PARAM4'] = SAREXT(low=df.low, 
                             high=df.high, 
                             startvalue= 100, 
                             offsetonreverse= .02, 
                             accelerationinitlong=0.02, 
                             accelerationlong=0.01, 
                             accelerationmaxlong=0.3, 
                             accelerationinitshort=0.03, 
                             accelerationshort=0.02, 
                             accelerationmaxshort=0.4)

df['date'] = pd.to_datetime(df.date)
df = df[['date', 'high', 'low', 'close', 'open', 'SAR', 'SAREXT', 'SAREXT_PARAM1', 'SAREXT_PARAM2', 'SAREXT_PARAM3', 'SAREXT_PARAM4']]
df.to_csv('spy_sarext.txt', index=False)
```

File: Indicators/ParabolicStopAndReverseExtended.cs [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+    /// <summary>
+    /// Parabolic SAR Extended Indicator 
+    /// Based on TA-Lib implementation
+    /// </summary>
+    public class ParabolicStopAndReverseExtended : BarIndicator, IIndicatorWarmUpPeriodProvider
+    {
+        private bool _isLong;
+        private IBaseDataBar _previousBar;
+        private decimal _sar;
+        private decimal _ep;
```
Comment (Martin-Molinero) [inline Indicators/ParabolicStopAndReverseExtended.cs:30]: `ep/sar/af/HasNegativeDM/etc` would be nice to expand these variable names to their full name, following the code style, makes it much easier to review/debug/understand 👍

File: Indicators/ParabolicStopAndReverseExtended.cs [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        /// <returns>A new value for this indicator</returns>
+        protected override decimal ComputeNextValue(IBaseDataBar input)
+        {
+            // On first iteration we can’t produce an SAR value so we save the current bar and return zero
+            if (Samples == 1){
+
+                 _previousBar = input;
+                // Makes sense to return _sarInit when its non-zero
+                if (_sarInit != 0) 
+                    return Math.Abs(_sarInit);
```
Comment (Martin-Molinero) [inline Indicators/ParabolicStopAndReverseExtended.cs:121]: - `){` following the project code style, opening brackets `{` are always in a new line & always present, even for 1 line `if` like this one. Please apply for all other other cases 👍 
- Why `Math.Abs()` in this case, given this extended version can return negative values? If this is still required though, suggest performing `Math.Abs()` once when this variable is set with a why comment explaining it 👍

File: Indicators/ParabolicStopAndReverseExtended.cs [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+                // Otherwise, return default
+                return input.Close; 
+            }
+
+            // On second iteration we initiate the position of extreme point SAR
+            if (Samples == 2)
+            {
+                Init(input);
+                _previousBar = input;
```
Comment (Martin-Molinero) [inline Indicators/ParabolicStopAndReverseExtended.cs:130]: Setting `_previousBar` before the calls to `Handle...` which will also use input as current bar? Should we be returning here?

File: Tests/Indicators/ParabolicStopAndReverseExtendedTests.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+        protected override string TestFileName => "spy_sarext.txt";
+
+        protected override string TestColumnName => "SAREXT"; 
+
+        [TestCase("SAREXT_PARAM1", 0.0, 0.0, 0.03, 0.02, 0.4, 0.02, 0.01, 0.3)]
+        [TestCase("SAREXT_PARAM2", 100, 0.0, 0.03, 0.02, 0.4, 0.02, 0.01, 0.3)]
+        [TestCase("SAREXT_PARAM3", -95, 0.0, 0.03, 0.02, 0.4, 0.02, 0.01, 0.3)]
+        [TestCase("SAREXT_PARAM4", 100, 0.02, 0.03, 0.02, 0.4, 0.02, 0.01, 0.3)]
+        public void ComparesWithExternalDataWithParams(string colName, decimal ss, decimal offset,
```
Comment (Martin-Molinero) [inline Tests/Indicators/ParabolicStopAndReverseExtendedTests.cs:41]: Nice! 👍

File: Indicators/ParabolicStopAndReverseExtended.cs [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+                }
+
+                // Calculate the new SAR
+                _sar = _sar + _afShort * (_ep - _sar);
+
+                // Make sure the new SAR is within yesterday's and today's range.
+                if (_sar < _previousBar.High)
+                    _sar = _previousBar.High;
+                if (_sar < currentBar.High)
+                    _sar = currentBar.High;
```
Comment (Martin-Molinero) [inline Indicators/ParabolicStopAndReverseExtended.cs:300]: These comparisons & assignments seem repeated a few times, seems they could be extracted into a private method to be reused? Will make the implementation easier to understand too I think

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Hey @nbkell! Thank you, leaving a few requests

Seems the original implementation intention was to match talib https://github.com/QuantConnect/Lean/pull/165/files so if we are now matching it correctly seems like a step in the good direction/bug fix 👍 

> I'm a bit confused on how @LouisSzeto test file is being generated, since when I run the same script locally I end up with a different .csv. For reference, I've posted below the script I'm using to generate the current version of spy_sarext.txt.

@LouisSzeto mind clarifying this, so we are sure the expected values are correct. Maybe it's a difference in the talib version being used? To standardize or avoid this issue, suggest using the `quantconnect/lean` image as base to generate the expected stats

**Comment by LouisSzeto** [comment]
@nbkell Hi! I actually use LeanCLI Research to generate the `df`. I looked at your code above and it is in the same logic that I created that, so no problem for the math (cc @Martin-Molinero).

**Comment by jaredbroad** [comment]
Closing as no response in 2 weeks. Please reopen if interested in picking it up.

**Comment by nbkell** [comment]
Apologies - I'm still planning to submit an updated PR but been bit busier lately and I haven't had the time. Will aim to submit within the next week. Thanks!

**Pull request closed without merging**